Prevent showing methods from blanket impls of not available foreign traits to show up in the search results

GuillaumeGomez opened the pull request:
Fixes https://github.com/rust-lang/rust/issues/115480.

In the case that the blanket impl trait is not available in the current crate, we prevent adding its methods in the search index.

Now how I found how to fix the issue: the `equivalent` method is not generated in the documentation pages but was still added to the search index. To render impls, we iterate over `cache.impls` so I took a look at how this was generated. Inside `formats/cache.rs`, we have `CacheBuilder::populate` where we push impls into `impls` but with this condition:

```rust
if cx.cache.traits.contains_key(&trait_did) {
```

I re-used this condition in `CacheBuilder::fold_item` to prevent this method from being added in `cache.search_index` or `cache.orphan_impl_items`.

PS: If you want to double-check if the added test works, just comment the code I added in `cache.rs` and it should fail.

r? @notriddle

GuillaumeGomez commented:
We already check for masked crate. Surprising. Let me check again...

GuillaumeGomez commented:
Also, `hashbrown` is not `doc(masked)`.

GuillaumeGomez commented:
It's not a private dependency either.

GuillaumeGomez commented:
Hum weird. I replaced my condition with a check on [trait_did.is_private_dep](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.is_private_dep) but it didn't work. Well, I'm going to investigate it more tomorrow.

GuillaumeGomez committed:
Prevent showing methods from blanket impls of not available foreign traits to show up in the search results

GuillaumeGomez committed:
Add regression test for #115480

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6480689942/job/17596703819) [(plain)](https://github.com/rust-lang/rust/commit/e39d96fd7344bcb3dd132b6a0b56a41e7a7278ba/checks/17596703819/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:8ade135a41bc03ea155e62e844d188df1ea18608)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_1d5267b9-3735-40d5-9e62-9d3c866eee75
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=foreign-blanket-impl
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_1d5267b9-3735-40d5-9e62-9d3c866eee75
GITHUB_REF=refs/pull/116597/merge
GITHUB_REF_NAME=116597/merge
GITHUB_REF_PROTECTED=false
---
Removing intermediate container 30d318825c4d
 ---> c6067be1063e
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> 3dac8fb88fec
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180117 sha256=2196c9034bf565528bbb1ee6dad4f753eb813f58822363e6b768f09c73e4d4ff
  Stored in directory: /tmp/pip-ephem-wheel-cache-w7hp14z9/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Can't uninstall 'setuptools'. No files were found to uninstall.
Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
Collecting virtualenv
  Downloading virtualenv-20.24.5-py3-none-any.whl (3.7 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.7/3.7 MB 66.8 MB/s eta 0:00:00
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.7-py2.py3-none-any.whl (468 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 104.5 MB/s eta 0:00:00
Collecting platformdirs<4,>=3.9.1
  Downloading platformdirs-3.11.0-py3-none-any.whl (17 kB)
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.12.4-py3-none-any.whl (11 kB)
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.7 filelock-3.12.4 platformdirs-3.11.0 virtualenv-20.24.5
Removing intermediate container f737655545db
 ---> e83ed9953b19
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> 8f16d5e33179
 ---> 8f16d5e33179
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> bcf16b8fd426
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container de61c5c0beec
 ---> b30dbe12658d
Successfully built b30dbe12658d
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:b30dbe12658d2c8fabcf28255fc80d9903ec0374d303ff0d88f8473a2db581bc
Uploading finished image sha256:b30dbe12658d2c8fabcf28255fc80d9903ec0374d303ff0d88f8473a2db581bc to https://ci-caches.rust-lang.org/docker/8849b25aebb63c7041ab10114da59fac9c6c89ff409673e53f6251b7e63c69daeaca7298d30885d05004ab27b231421908523f297222d07a53450f37e4691d72
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
b30dbe12658d   1 second ago     /bin/sh -c #(nop)  ENV SCRIPT=TIDY_PRINT_DIF…   0B        
8f16d5e33179   2 seconds ago    /bin/sh -c #(nop) COPY file:078ea1d11e7b7cda…   367B      
e83ed9953b19   3 seconds ago    |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.9MB    
3dac8fb88fec   10 seconds ago   /bin/sh -c #(nop) COPY file:ac591dd6bc5afa66…   5.33kB    
c6067be1063e   11 seconds ago   |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.1MB    
---
<missing>      2 weeks ago      /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      2 weeks ago      /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      2 weeks ago      /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7fe6c494f750>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/8849b25aebb63c7041ab10114da59fac9c6c89ff409673e53f6251b7e63c69daeaca7298d30885d05004ab27b231421908523f297222d07a53450f37e4691d72
sha256:b30dbe12658d2c8fabcf28255fc80d9903ec0374d303ff0d88f8473a2db581bc
---
DirectMap4k:      251840 kB
DirectMap2M:     8136704 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/6d05c430d2f0573bf98320bb1d2457a72ea1e67f/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-6d05c430d2f0573bf98320bb1d2457a72ea1e67f-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 29.34s
##[endgroup]
fmt check
##[error]Diff in /checkout/src/librustdoc/formats/cache.rs at line 316:
                                     // If this is a foreign trait impl but the trait documentation
                                     // is not available, we should not allow the methods to show up
                                     // in the search results.
-                                    if !trait_did.is_local() && self.tcx.is_private_dep(trait_did.krate)
+                                    if !trait_did.is_local()
+                                        && self.tcx.is_private_dep(trait_did.krate)
                                         return None;
                                     }
                                     }
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/src/librustdoc/clean/inline.rs" "/checkout/src/librustdoc/clean/utils.rs" "/checkout/src/librustdoc/clean/auto_trait.rs" "/checkout/src/librustdoc/formats/item_type.rs" "/checkout/src/librustdoc/clean/types/tests.rs" "/checkout/src/librustdoc/clean/types.rs" "/checkout/src/librustdoc/formats/cache.rs" "/checkout/src/librustdoc/clean/render_macro_matchers.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Wed Oct 11 09:39:30 UTC 2023
  network time: Wed, 11 Oct 2023 09:39:30 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

GuillaumeGomez committed:
Handle private dep at the same level as masked crates

GuillaumeGomez commented:
Fixed. The problem was that the wrong `search-index.js` file was picked in the tests so once I removed the whole folder, it worked as expected. I also updated the test to ensure this is a private dep and it's being imported.

bors commented:
:pushpin: Commit 2d37b00e24bfb1051e4a1d587bf766378d19e9b7 has been approved by `notriddle`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle 2d37b00e24bfb1051e4a1d587bf766378d19e9b7 -->
<!-- homu: {"type":"Approved","sha":"2d37b00e24bfb1051e4a1d587bf766378d19e9b7","approver":"notriddle","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
